File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int hBT(TreeNode* root){
        if(root==NULL) return 0;
        int left = hBT(root->left);
        int right = hBT(root->right);
        return max(left,right)+1;
    }
    int diameter(TreeNode* root){
        if(root==NULL) return 0;
        int op1 = diameter(root->left);
        int op2 = diameter(root->right);
        int op3 = hBT(root->left)+hBT(root->right);
        int ans = max(op1,max(op2,op3));
        return ans;
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        return diameter(root);
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

let height = (root)=>{
    if(root==null) return 0;
    let left = height(root.left);
    let right = height(root.right);
    return Math.max(left,right)+1;
}
var diameterOfBinaryTree = function(root) {
    if(root==null) return 0;
    let left = diameterOfBinaryTree(root.left);
    let right = diameterOfBinaryTree(root.right);
    let heightX = height(root.left)+height(root.right);
    return Math.max(left,Math.max(right,heightX));
};